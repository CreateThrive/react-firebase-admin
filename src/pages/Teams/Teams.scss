.media-list {
  margin-left: -1.5rem;
  margin-right: -1.5rem;
}

.media {
  padding: 1rem 1.5rem;
}

.button {
  display: inline-flex;
  align-self: center;
  margin-right: 0.75rem;
}
