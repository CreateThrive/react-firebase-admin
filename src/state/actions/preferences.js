import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const setUserLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
