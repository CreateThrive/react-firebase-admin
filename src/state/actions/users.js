import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';

import { firebaseError, deleteLogo, getLogoUrl, uploadLogo } from 'utils';
import firebase from 'firebase.js';
import { checkUserData, AUTH_UPDATE_USER_DATA } from './auth';
import {
  fetchCollection,
  fetchDocument,
  createDocument,
  deleteDocument,
  updateDocument,
} from '../api';

export const USERS_FETCH_DATA_INIT = createAction('USERS_FETCH_DATA_INIT');
export const USERS_FETCH_DATA_SUCCESS = createAction(
  'USERS_FETCH_DATA_SUCCESS'
);
export const USERS_FETCH_DATA_FAIL = createAction('USERS_FETCH_DATA_FAIL');

export const USERS_DELETE_USER_INIT = createAction('USERS_DELETE_USER_INIT');
export const USERS_DELETE_USER_SUCCESS = createAction(
  'USERS_DELETE_USER_SUCCESS'
);
export const USERS_DELETE_USER_FAIL = createAction('USERS_DELETE_USER_FAIL');

export const USERS_CREATE_USER_INIT = createAction('USERS_CREATE_USER_INIT');
export const USERS_CREATE_USER_SUCCESS = createAction(
  'USERS_CREATE_USER_SUCCESS'
);
export const USERS_CREATE_USER_FAIL = createAction('USERS_CREATE_USER_FAIL');

export const USERS_MODIFY_USER_INIT = createAction('USERS_MODIFY_USER_INIT');
export const USERS_MODIFY_USER_SUCCESS = createAction(
  'USERS_MODIFY_USER_SUCCESS'
);
export const USERS_MODIFY_USER_FAIL = createAction('USERS_MODIFY_USER_FAIL');

export const USERS_CLEAN_UP = createAction('USERS_CLEAN_UP');

export const USERS_CLEAR_DATA_LOGOUT = createAction('USERS_CLEAR_DATA_LOGOUT');

export const USERS_ADD_USER = createAction('USERS_ADD_USER');

export const USERS_REMOVE_USER = createAction('USERS_REMOVE_USER');

export const fetchUsers = (userId = '') => {
  return async (dispatch, getState) => {
    dispatch(checkUserData());
    const { locale } = getState().preferences;
    const { id } = getState().auth.userData;

    dispatch(USERS_FETCH_DATA_INIT());

    if (userId) {
      let user;
      try {
        user = await fetchDocument('users', userId);
      } catch (error) {
        toastr.error('', error);
        return dispatch(USERS_FETCH_DATA_FAIL({ error }));
      }

      if (!user) {
        const errorMessage = firebaseError('users.UserNotAvailable', locale);
        toastr.error('', errorMessage);
        return dispatch(USERS_FETCH_DATA_FAIL({ error: errorMessage }));
      }

      const users = getState().users.data;
      users.push(user);

      return dispatch(
        USERS_FETCH_DATA_SUCCESS({
          data: users,
        })
      );
    }

    let users;

    try {
      users = await fetchCollection('users');
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(USERS_FETCH_DATA_FAIL({ error }));
    }

    return dispatch(
      USERS_FETCH_DATA_SUCCESS({
        data: users.filter((user) => user.id !== id),
      })
    );
  };
};

export const deleteUser = (id) => {
  return async (dispatch, getState) => {
    dispatch(USERS_DELETE_USER_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState()
      .users.data.filter((user) => user.id === id)
      .pop();

    const deleteLogoTask = logoUrl ? deleteLogo(logoUrl, 'users') : null;

    const deleteUserTask = deleteDocument('users', id);

    try {
      await Promise.all([deleteLogoTask, deleteUserTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_DELETE_USER_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'The user was deleted.');
    return dispatch(USERS_DELETE_USER_SUCCESS({ id }));
  };
};

export const clearUsersDataLogout = () => {
  return (dispatch) => {
    dispatch(USERS_CLEAR_DATA_LOGOUT());
  };
};

export const createUser = ({
  name,
  email,
  location,
  file,
  createdAt,
  isAdmin,
}) => {
  return async (dispatch, getState) => {
    dispatch(USERS_CREATE_USER_INIT());
    const { locale } = getState().preferences;

    let response;
    try {
      const createUserAuth = firebase
        .functions()
        .httpsCallable('httpsCreateUser');

      response = await createUserAuth({ email, isAdmin });
    } catch (error) {
      const errorMessage = firebaseError(error.message, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_CREATE_USER_FAIL({
          error: errorMessage,
        })
      );
    }

    const { uid } = response.data;

    let uploadLogoTask = null;
    let logoUrl = null;
    if (file) {
      logoUrl = getLogoUrl(uid, file, 'users');
      uploadLogoTask = uploadLogo(uid, file, 'users');
    }
    const userData = {
      name,
      email,
      location,
      logoUrl,
      createdAt,
      isAdmin,
      teams: [],
    };

    const createUserDbTask = createDocument('users', uid, userData);

    const actionCodeSettings = {
      url: process.env.REACT_APP_LOGIN_PAGE_URL,
      handleCodeInApp: true,
    };

    const sendSignInLinkToEmailTask = firebase
      .auth()
      .sendSignInLinkToEmail(email, actionCodeSettings);

    try {
      await Promise.all([
        uploadLogoTask,
        createUserDbTask,
        sendSignInLinkToEmailTask,
      ]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_CREATE_USER_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'User created successfully');
    return dispatch(USERS_CREATE_USER_SUCCESS({ user: response.data }));
  };
};

export const modifyUser = ({
  name,
  location,
  isAdmin,
  file,
  createdAt,
  id,
  isEditing,
  isProfile,
}) => {
  return async (dispatch, getState) => {
    dispatch(USERS_MODIFY_USER_INIT());
    const { locale } = getState().preferences;
    const user = isProfile
      ? getState().auth.userData
      : getState().users.data.find((thisUser) => thisUser.id === id);
    const { logoUrl } = user;
    let deleteLogoTask;
    let uploadLogoTask;
    let newLogoUrl = null;
    if (file) {
      newLogoUrl = getLogoUrl(id, file, 'users');
      deleteLogoTask = logoUrl && deleteLogo(logoUrl, 'users');
      uploadLogoTask = uploadLogo(id, file, 'users');
    }

    const userData = {
      name,
      location,
      createdAt,
      isAdmin: isAdmin || user.isAdmin,
      logoUrl: logoUrl || newLogoUrl,
    };
    const updateUserDbTask = updateDocument('users', id, userData);

    try {
      await Promise.all([deleteLogoTask, uploadLogoTask, updateUserDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        USERS_MODIFY_USER_FAIL({
          error: errorMessage,
        })
      );
    }

    if (isProfile) {
      dispatch(AUTH_UPDATE_USER_DATA({ ...userData, id }));
    }

    if (isProfile) {
      toastr.success('', 'Profile updated successfully');
    } else if (isEditing) {
      toastr.success('', 'User updated successfully');
    }

    return dispatch(USERS_MODIFY_USER_SUCCESS({ user: userData, id }));
  };
};

export const usersCleanUp = () => (dispatch) => dispatch(USERS_CLEAN_UP());

export const removeStoreUser = (userId) => (dispatch) =>
  dispatch(USERS_REMOVE_USER({ userId }));

export const addStoreUser = (user) => (dispatch) =>
  dispatch(USERS_ADD_USER({ user }));
