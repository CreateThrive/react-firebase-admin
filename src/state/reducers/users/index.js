import { createReducer } from 'redux-act';

import {
  USERS_FETCH_DATA_INIT,
  USERS_FETCH_DATA_SUCCESS,
  USERS_FETCH_DATA_FAIL,
  USERS_DELETE_USER_INIT,
  USERS_DELETE_USER_SUCCESS,
  USERS_DELETE_USER_FAIL,
  USERS_CLEAR_DATA,
  USERS_CREATE_USER_INIT,
  USERS_CREATE_USER_SUCCESS,
  USERS_CREATE_USER_FAIL,
  USERS_MODIFY_USER_INIT,
  USERS_MODIFY_USER_SUCCESS,
  USERS_MODIFY_USER_FAIL,
  USERS_CLEAN_UP,
  USERS_CLEAR_DATA_LOGOUT
} from 'state/actions/users';

const initialState = {
  data: [],
  loading: false,
  error: null,
  success: false,
  deleted: false
};

export const usersReducer = createReducer(
  {
    [USERS_FETCH_DATA_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [USERS_FETCH_DATA_SUCCESS]: (state, payload) => ({
      ...state,
      data: payload.users,
      loading: false,
      error: null
    }),
    [USERS_FETCH_DATA_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [USERS_DELETE_USER_INIT]: state => ({
      ...state,
      loading: true
    }),
    [USERS_DELETE_USER_SUCCESS]: (state, payload) => ({
      ...state,
      data: state.data.filter(elem => elem.id !== payload.id),
      loading: false,
      error: null,
      deleted: true
    }),
    [USERS_DELETE_USER_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [USERS_CLEAR_DATA]: state => ({
      ...initialState,
      data: state.data
    }),
    [USERS_CREATE_USER_INIT]: state => ({
      ...state,
      loading: true
    }),
    [USERS_CREATE_USER_SUCCESS]: (state, payload) => ({
      ...state,
      data: state.data.concat(payload.user),
      loading: false,
      error: null,
      success: true
    }),
    [USERS_CREATE_USER_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [USERS_MODIFY_USER_INIT]: state => ({
      ...state,
      loading: true
    }),
    [USERS_MODIFY_USER_SUCCESS]: (state, payload) => ({
      ...state,
      data: state.data.map(elem => {
        if (elem.id === payload.id) {
          return {
            name: payload.user.name,
            location: payload.user.location,
            id: payload.id,
            logoUrl: payload.user.logoUrl,
            createdAt: payload.user.createdAt,
            email: elem.email
          };
        }
        return elem;
      }),
      loading: false,
      error: null,
      success: true
    }),
    [USERS_MODIFY_USER_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [USERS_CLEAN_UP]: state => ({
      ...state,
      loading: false,
      error: null,
      success: false,
      deleted: false
    }),
    [USERS_CLEAR_DATA_LOGOUT]: () => ({
      ...initialState
    })
  },
  initialState
);
