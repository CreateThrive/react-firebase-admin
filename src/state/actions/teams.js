import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';
import { v4 as uuid } from 'uuid';

import { firebaseError, deleteLogo, getLogoUrl, uploadLogo } from 'utils';
import {
  fetchCollection,
  createDocument,
  deleteDocument,
  updateDocument,
  batchUpdateDocument,
} from '../api';
import { removeStoreUser, addStoreUser } from './users';

export const TEAMS_FETCH_TEAMS_INIT = createAction('TEAMS_FETCH_TEAMS_INIT');

export const TEAMS_FETCH_TEAMS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAMS_SUCCESS'
);

export const TEAMS_FETCH_TEAMS_FAIL = createAction('TEAMS_FETCH_TEAMS_FAIL');

export const TEAMS_DELETE_TEAM_INIT = createAction('TEAMS_DELETE_TEAM_INIT');

export const TEAMS_DELETE_TEAM_SUCCESS = createAction(
  'TEAMS_DELETE_TEAM_SUCCESS'
);

export const TEAMS_DELETE_TEAM_FAIL = createAction('TEAMS_DELETE_TEAM_FAIL');

export const TEAMS_CREATE_TEAM_INIT = createAction('TEAMS_CREATE_TEAM_INIT');

export const TEAMS_CREATE_TEAM_SUCCESS = createAction(
  'TEAMS_CREATE_TEAM_SUCCESS'
);
export const TEAMS_CREATE_TEAM_FAIL = createAction('TEAMS_CREATE_TEAM_FAIL');

export const TEAMS_MODIFY_TEAM_INIT = createAction('TEAMS_MODIFY_TEAM_INIT');

export const TEAMS_MODIFY_TEAM_SUCCESS = createAction(
  'TEAMS_MODIFY_TEAM_SUCCESS'
);
export const TEAMS_MODIFY_TEAM_FAIL = createAction('TEAMS_MODIFY_TEAM_FAIL');

export const TEAMS_CLEAN_UP = createAction('TEAMS_CLEAN_UP');

export const TEAMS_ADD_USER_INIT = createAction('TEAMS_ADD_USER_INIT');

export const TEAMS_ADD_USER_SUCCESS = createAction('TEAMS_ADD_USER_SUCCESS');

export const TEAMS_ADD_USER_FAIL = createAction('TEAMS_ADD_USER_FAIL');

export const TEAMS_REMOVE_USER_INIT = createAction('TEAMS_REMOVE_USER_INIT');

export const TEAMS_REMOVE_USER_SUCCESS = createAction(
  'TEAMS_REMOVE_USER_SUCCESS'
);

export const TEAMS_REMOVE_USER_FAIL = createAction('TEAMS_REMOVE_USER_FAIL');

export const TEAMS_SELECTED_TEAM = createAction('TEAMS_SELECTED_TEAM');

export const TEAMS_FETCH_TEAM_USERS_INIT = createAction(
  'TEAMS_FETCH_TEAM_USERS_INIT'
);

export const TEAMS_FETCH_TEAM_USERS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAM_USERS_SUCCESS'
);

export const TEAMS_FETCH_TEAM_USERS_FAIL = createAction(
  'TEAMS_FETCH_TEAM_USERS_FAIL'
);

export const fetchTeams = () => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAMS_INIT());
    const { locale } = getState().preferences;
    const { isAdmin, id } = getState().auth.userData;

    let teams;
    try {
      teams = await fetchCollection(
        'teams',
        isAdmin
          ? {
              filterBy: 'users',
              value: id,
              isArray: true,
            }
          : {}
      );
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAMS_FAIL({ error }));
    }

    return dispatch(
      TEAMS_FETCH_TEAMS_SUCCESS({
        teams: isAdmin
          ? teams
          : teams.filter((value) => value.users?.includes(id)),
      })
    );
  };
};

export const deleteTeam = (id) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_DELETE_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    const deleteLogoTask = logoUrl ? deleteLogo(logoUrl, 'teams') : null;

    const deleteTeamTask = deleteDocument('teams', id);

    try {
      await Promise.all([deleteLogoTask, deleteTeamTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_DELETE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'The team was deleted.');
    return dispatch(TEAMS_DELETE_TEAM_SUCCESS({ id }));
  };
};

export const createTeam = ({ name, description, file }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_CREATE_TEAM_INIT());
    const { locale } = getState().preferences;

    const uid = uuid();

    let uploadLogoTask = null;
    let logoUrl = null;
    if (file) {
      logoUrl = getLogoUrl(uid, file, 'teams');
      uploadLogoTask = uploadLogo(uid, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl,
      createdAt: new Date().toString(),
      createdBy: getState().auth.userData.name,
    };

    const createTeamDbTask = createDocument('teams', uid, teamData);

    try {
      await Promise.all([uploadLogoTask, createTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_CREATE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team created successfully');
    return dispatch(
      TEAMS_CREATE_TEAM_SUCCESS({ team: { ...teamData, id: uid } })
    );
  };
};

export const modifyTeam = ({ name, description, file, id }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_MODIFY_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    let deleteLogoTask;
    let uploadLogoTask;
    let newLogoUrl = null;
    if (file) {
      newLogoUrl = getLogoUrl(id, file, 'teams');
      deleteLogoTask = logoUrl && deleteLogo(logoUrl, 'teams');
      uploadLogoTask = uploadLogo(id, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl: logoUrl || newLogoUrl,
    };
    const updateTeamDbTask = updateDocument('teams', id, teamData);

    try {
      await Promise.all([deleteLogoTask, uploadLogoTask, updateTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_MODIFY_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team updated successfully');

    return dispatch(TEAMS_MODIFY_TEAM_SUCCESS({ team: teamData, id }));
  };
};

export const teamsCleanUp = () => (dispatch) => dispatch(TEAMS_CLEAN_UP());

export const addUser = (teamId, user) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_ADD_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList } = getState().teams;

    const userId = user.id;

    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = [...userTeams, teamId];
    const newTeamUsers = [...teamUsers, userId];

    const batchCollections = {
      users: {
        documents: [
          {
            id: userId,
            field: 'teams',
            value: newUserTeams,
          },
        ],
      },

      teams: {
        documents: [
          {
            id: teamId,
            field: 'users',
            value: newTeamUsers,
          },
        ],
      },
    };

    try {
      await batchUpdateDocument(batchCollections);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_ADD_USER_FAIL({ error }));
    }

    dispatch(removeStoreUser(userId));

    return dispatch(
      TEAMS_ADD_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        user: { ...user, teams: newUserTeams },
      })
    );
  };
};

export const removeUser = (teamId, userId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_REMOVE_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList, usersList } = getState().teams;

    const user = usersList.find((value) => value.id === userId);
    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = userTeams.filter((value) => value !== teamId);
    const newTeamUsers = teamUsers.filter((value) => value !== userId);

    const bulkUpdate = {};
    bulkUpdate[`/users/${userId}/teams`] = newUserTeams;
    bulkUpdate[`/teams/${teamId}/users`] = newTeamUsers;

    try {
      await updateDocument('', '', bulkUpdate);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_REMOVE_USER_FAIL({ error }));
    }

    dispatch(addStoreUser({ ...user, teams: newUserTeams }));

    return dispatch(
      TEAMS_REMOVE_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        userId,
      })
    );
  };
};

export const fetchTeamUsers = (teamId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAM_USERS_INIT());
    const { locale } = getState().preferences;

    let users;
    try {
      users = await fetchCollection('users', {
        filterBy: 'users',
        value: teamId,
        isArray: true,
      });
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAM_USERS_FAIL({ error }));
    }

    dispatch(TEAMS_SELECTED_TEAM({ teamId }));

    return dispatch(
      TEAMS_FETCH_TEAM_USERS_SUCCESS({
        users: users.filter((value) => value.teams?.includes(teamId)),
      })
    );
  };
};
