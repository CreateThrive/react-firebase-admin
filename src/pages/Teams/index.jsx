import React, { useEffect } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';

import { useFormatMessage } from 'hooks';
import TeamsCard from 'components/TeamsCard';
import UsersCard from 'components/UsersCard';
import { fetchTeamUsers } from 'state/actions/teams';

import './Teams.scss';

const Teams = () => {
  const { teamId, teamsList } = useSelector(
    (state) => ({
      teamsList: state.teams.teamsList,
      teamId: state.teams.teamId,
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  useEffect(() => {
    if (!teamId && teamsList.length > 0) {
      const newTeamId = teamsList[0].id;
      dispatch(fetchTeamUsers(newTeamId));
    }
  }, [dispatch, teamId, teamsList]);

  return (
    <>
      <section className="hero is-hero-bar">
        <div className="hero-body">
          <div className="level">
            <div className="level-left">
              <div className="level-item">
                <h1 className="title">{useFormatMessage('Teams.teams')}</h1>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section className="section is-main-section">
        <div className="columns">
          <div className="column">
            <TeamsCard />
          </div>
          <div className="column">
            <UsersCard />
          </div>
        </div>
      </section>
    </>
  );
};

export default Teams;
