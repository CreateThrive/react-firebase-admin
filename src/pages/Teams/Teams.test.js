import React from 'react';
import * as reactRedux from 'react-redux';

import * as teamsActions from 'state/actions/teams';
import * as usersActions from 'state/actions/users';
import Teams from '.';

describe('<Teams /> rendering', () => {
  const dispatchMock = jest.fn();
  const state = {
    auth: { userData: { isAdmin: true } },
    teams: {
      success: false,
      deleted: false,
      loadingUsers: false,
      loadingTeams: false,
      teamsList: [{ id: 'teamTestId', name: 'testTeam', createdBy: 'test' }],
      usersList: [
        {
          name: 'testUser',
          id: 'userTestId',
          isAdmin: false,
          email: 'test@test.com',
          location: 'Montevideo, Uruguay',
          file: null,
          logoUrl: 'some lougoUrl',
          createdAt: '11/12/2020',
        },
      ],
    },
    users: { data: ['testUser'] },
  };

  beforeEach(() => {
    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => dispatchMock);
    jest.spyOn(teamsActions, 'teamsCleanUp').mockImplementation(jest.fn);
    jest.spyOn(usersActions, 'usersCleanUp').mockImplementation(jest.fn);
  });

  it('should render without crashing', () => {
    const { component } = renderWithProviders(<Teams />)({ ...state });

    expect(component.asFragment()).toMatchSnapshot();
  });
});
