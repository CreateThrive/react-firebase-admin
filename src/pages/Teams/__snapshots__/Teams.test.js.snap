// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Teams /> rendering should render without crashing 1`] = `
<DocumentFragment>
  <section
    class="hero is-hero-bar"
  >
    <div
      class="hero-body"
    >
      <div
        class="level"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          >
            <h1
              class="title"
            >
              Teams
            </h1>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section
    class="section is-main-section"
  >
    <div
      class="columns"
    >
      <div
        class="column"
      >
        <div
          class="card has-height-medium"
        >
          <header
            class="card-header"
          >
            <p
              class="card-header-title"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-account-group default"
                />
              </span>
              <span>
                Available teams
              </span>
            </p>
            <button
              class="button"
              type="button"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-account-multiple-plus default"
                />
              </span>
              <span>
                Add Team
              </span>
            </button>
          </header>
          <div>
            <div
              class="field-body"
              style="padding: 0.5rem 0.25rem;"
            >
              <div
                class="field"
              >
                <div
                  class="control is-expanded"
                >
                  <input
                    class="input search"
                    placeholder="Search by name"
                    style="height: 38px;"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="field"
              >
                <div
                  class="control is-expanded"
                >
                  <div
                    style="min-width: 220px;"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="card-content ps ps--active-y"
            style="padding: 0px 1.5rem; overflow: auto; max-height: 700px;"
          >
            <div
              class="media-list"
            >
              <div
                class="media"
                data-testid="media"
                style="margin-top: auto; margin-bottom: auto; padding-bottom: 1rem; cursor: pointer;"
              >
                <figure
                  class="media-left"
                >
                  <p
                    class="image is-64x64"
                  >
                    <img
                      alt="Profile logo preview"
                      class="is-rounded"
                      src="team-default-log.png"
                    />
                  </p>
                </figure>
                <div
                  class="media-content"
                >
                  <div
                    class="content"
                  >
                    <p
                      class="media-meta"
                    >
                      <strong>
                        testTeam
                      </strong>
                       
                    </p>
                    <p />
                  </div>
                </div>
                <div
                  class="media-right"
                >
                  <div
                    class="field is-grouped is-grouped-multiline"
                  >
                    <div
                      class="control"
                    >
                      <span
                        class="icon has-text-info"
                        data-testid="modify"
                        style="cursor: pointer;"
                      >
                        <i
                          class="mdi mdi-pencil"
                        />
                      </span>
                    </div>
                    <div
                      class="control"
                    >
                      <span
                        class="icon has-text-danger"
                        data-testid="delete"
                        style="cursor: pointer;"
                      >
                        <i
                          class="mdi mdi-cancel"
                        />
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="column"
      >
        <div
          class="card has-height-medium"
        >
          <header
            class="card-header"
          >
            <p
              class="card-header-title"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-account-supervisor default"
                />
              </span>
              <span>
                Team users
              </span>
            </p>
          </header>
          <div>
            <div
              class="field-body"
              style="padding: 0.5rem 0.25rem;"
            >
              <div
                class="field"
              >
                <div
                  class="control is-expanded"
                >
                  <input
                    class="input search"
                    placeholder="Search by name"
                    style="height: 38px;"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="field"
              >
                <div
                  class="control is-expanded"
                >
                  <div
                    style="min-width: 220px;"
                  >
                    <div
                      class=" css-2b097c-container"
                    >
                      <div
                        class=" css-spersy-control"
                      >
                        <div
                          class=" css-g1d714-ValueContainer"
                        >
                          <div
                            class=" css-1wa3eu0-placeholder"
                          >
                            Add User
                          </div>
                          <div
                            class="css-b8ldur-Input"
                          >
                            <div
                              class=""
                              style="display: inline-block;"
                            >
                              <input
                                aria-autocomplete="list"
                                autocapitalize="none"
                                autocomplete="off"
                                autocorrect="off"
                                id="react-select-2-input"
                                spellcheck="false"
                                style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                                tabindex="0"
                                type="text"
                                value=""
                              />
                              <div
                                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                              />
                            </div>
                          </div>
                        </div>
                        <div
                          class=" css-1hb7zxy-IndicatorsContainer"
                        >
                          <span
                            class=" css-1okebmr-indicatorSeparator"
                          />
                          <div
                            aria-hidden="true"
                            class=" css-tlfecz-indicatorContainer"
                          >
                            <svg
                              aria-hidden="true"
                              class="css-6q0nyr-Svg"
                              focusable="false"
                              height="20"
                              viewBox="0 0 20 20"
                              width="20"
                            >
                              <path
                                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="card-content ps ps--active-y"
            style="padding: 0px 1.5rem; overflow: auto; max-height: 700px;"
          >
            <div
              class="media-list"
            >
              <div
                class="media"
                data-testid="media"
                style="margin-top: auto; margin-bottom: auto; padding-bottom: 1rem;"
              >
                <figure
                  class="media-left"
                >
                  <p
                    class="image is-64x64"
                  >
                    <img
                      alt="Profile logo preview"
                      class="is-rounded"
                      src="some lougoUrl"
                    />
                  </p>
                </figure>
                <div
                  class="media-content"
                >
                  <div
                    class="content"
                  >
                    <p
                      class="media-meta"
                    >
                      <strong>
                        testUser
                      </strong>
                       
                      <small
                        class="has-text-grey"
                      >
                        Thu, Nov 12, 2020
                      </small>
                    </p>
                    <p>
                      test@test.com
                    </p>
                  </div>
                </div>
                <div
                  class="media-right"
                >
                  <div
                    class="field is-grouped is-grouped-multiline"
                  >
                    <div
                      class="control"
                    >
                      <a
                        href="/users/userTestId"
                      >
                        <span
                          class="icon"
                        >
                          <i
                            class="mdi mdi-account-edit"
                          />
                        </span>
                      </a>
                    </div>
                    <div
                      class="control"
                    >
                      <span
                        class="icon has-text-danger"
                        data-testid="delete"
                        style="cursor: pointer;"
                      >
                        <i
                          class="mdi mdi-cancel"
                        />
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
